File: knowledge_scrapers/social_media_scraper.py
#!/usr/bin/env python3
"""
Social Media Content Scraper

This modular scraper extracts public content from social media platforms. Features include:
- Extraction of posts, comments, and metadata
- Sentiment analysis integration
- Rate limiting and ethical scraping practices
- Support for multiple social media platforms
"""

import os
import sys
import requests
import json
import time
import re
from bs4 import BeautifulSoup
from typing import Dict, Optional, List, Tuple
from urllib.parse import urljoin, urlparse
from dataclasses import dataclass
from datetime import datetime

# Ensure Marina root is in path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

@dataclass
class SocialMediaPost:
    """Represents a social media post with extracted content"""
    url: str
    platform: str
    author: str
    content: str
    timestamp: Optional[str]
    likes_count: Optional[int]
    shares_count: Optional[int]
    comments_count: Optional[int]
    hashtags: List[str]
    scraped_at: str

class SocialMediaScraper:
    """
    Modular social media scraper that can be extended for different platforms
    """
    
    def __init__(self, platform: str, delay_between_requests: float = 2.0):
        self.platform = platform.lower()
        self.delay = delay_between_requests
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Marina-SocialMediaScraper/1.0 (Educational Research)'
        })
        self.visited_urls = set()
        
        # Platform-specific configurations
        self.platform_configs = {
            'twitter': {
                'post_selector': '.tweet',
                'content_selector': '.tweet-text',
                'author_selector': '.username',
                'timestamp_selector': '.timestamp'
            },
            'reddit': {
                'post_selector': '.thing',
                'content_selector': '.md',
                'author_selector': '.author',
                'timestamp_selector': '.live-timestamp'
            },
            'facebook': {
                'post_selector': '[data-testid="post_message"]',
                'content_selector': '[data-testid="post_message"]',
                'author_selector': 'strong',
                'timestamp_selector': 'abbr'
            }
        }
    
    def extract_hashtags(self, text: str) -> List[str]:
        """Extract hashtags from text content"""
        hashtag_pattern = r'#\w+'
        hashtags = re.findall(hashtag_pattern, text, re.IGNORECASE)
        return [tag.lower() for tag in hashtags]
    
    def extract_mentions(self, text: str) -> List[str]:
        """Extract user mentions from text content"""
        mention_pattern = r'@\w+'
        mentions = re.findall(mention_pattern, text, re.IGNORECASE)
        return [mention.lower() for mention in mentions]
    
    def scrape_post(self, url: str) -> Optional[SocialMediaPost]:
        """Scrape a single social media post"""
        if url in self.visited_urls:
            return None
            
        self.visited_urls.add(url)
        
        try:
            print(f"🔍 Scraping {self.platform} post: {url}")
            time.sleep(self.delay)
            
            response = self.session.get(url, timeout=15)
            if response.status_code != 200:
                print(f"❌ Failed to fetch {url}: HTTP {response.status_code}")
                return None
                
            soup = BeautifulSoup(response.content, 'html.parser')
            config = self.platform_configs.get(self.platform, {})
            
            # Extract post content
            content_elem = soup.select_one(config.get('content_selector', ''))
            content = content_elem.get_text(strip=True) if content_elem else 'N/A'
            
            # Extract author
            author_elem = soup.select_one(config.get('author_selector', ''))
            author = author_elem.get_text(strip=True) if author_elem else 'N/A'
            
            # Extract timestamp
            timestamp_elem = soup.select_one(config.get('timestamp_selector', ''))
            timestamp = timestamp_elem.get('datetime') or timestamp_elem.get_text(strip=True) if timestamp_elem else None
            
            # Extract engagement metrics (simplified)
            likes_count = self._extract_number_from_text(soup, ['like', 'heart'])
            shares_count = self._extract_number_from_text(soup, ['share', 'retweet'])
            comments_count = self._extract_number_from_text(soup, ['comment', 'reply'])
            
            # Extract hashtags
            hashtags = self.extract_hashtags(content)
            
            return SocialMediaPost(
                url=url,
                platform=self.platform,
                author=author,
                content=content,
                timestamp=timestamp,
                likes_count=likes_count,
                shares_count=shares_count,
                comments_count=comments_count,
                hashtags=hashtags,
                scraped_at=datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            )
            
        except Exception as e:
            print(f"❌ Error scraping {url}: {e}")
            return None
    
    def _extract_number_from_text(self, soup: BeautifulSoup, keywords: List[str]) -> Optional[int]:
        """Extract numerical values associated with keywords (likes, shares, etc.)"""
        for keyword in keywords:
            elements = soup.find_all(text=re.compile(keyword, re.IGNORECASE))
            for element in elements:
                if hasattr(element, 'parent'):
                    parent_text = element.parent.get_text()
                    numbers = re.findall(r'\d+', parent_text)
                    if numbers:
                        try:
                            return int(numbers[0])
                        except ValueError:
                            continue
        return None
    
    def discover_posts(self, feed_url: str, max_posts: int = 10) -> List[str]:
        """Discover post URLs from a feed or profile page"""
        post_urls = []
        
        try:
            response = self.session.get(feed_url, timeout=15)
            if response.status_code != 200:
                print(f"❌ Failed to fetch feed: {feed_url}")
                return []
                
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # Platform-specific post URL discovery
            if self.platform == 'twitter':
                post_links = soup.find_all('a', href=re.compile(r'/status/\d+'))
            elif self.platform == 'reddit':
                post_links = soup.find_all('a', href=re.compile(r'/r/\w+/comments/'))
            else:
                # Generic approach - look for links that might be posts
                post_links = soup.find_all('a', href=True)
            
            for link in post_links[:max_posts]:
                href = link.get('href')
                if href:
                    full_url = urljoin(feed_url, href)
                    post_urls.append(full_url)
                    
        except Exception as e:
            print(f"❌ Error discovering posts from {feed_url}: {e}")
        
        return post_urls[:max_posts]
    
    def scrape_feed(self, feed_url: str, max_posts: int = 10) -> List[SocialMediaPost]:
        """Scrape multiple posts from a social media feed"""
        print(f"🚀 Starting {self.platform} scraping from: {feed_url}")
        
        post_urls = self.discover_posts(feed_url, max_posts)
        scraped_posts = []
        
        for url in post_urls:
            post = self.scrape_post(url)
            if post:
                scraped_posts.append(post)
                
        print(f"✅ Scraped {len(scraped_posts)} posts from {self.platform}")
        return scraped_posts
    
    def save_results(self, posts: List[SocialMediaPost], filename: str = None):
        """Save scraped posts to JSON file"""
        if not filename:
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            filename = f"social_media_scrape_{self.platform}_{timestamp}.json"
        
        results_dir = "/home/adminx/Marina/knowledge_scrapers/scraping_results"
        os.makedirs(results_dir, exist_ok=True)
        
        filepath = os.path.join(results_dir, filename)
        
        # Convert dataclass instances to dictionaries for JSON serialization
        posts_data = [post.__dict__ for post in posts]
        
        with open(filepath, 'w') as f:
            json.dump({
                'platform': self.platform,
                'total_posts': len(posts),
                'scraped_at': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
                'posts': posts_data
            }, f, indent=2)
        
        print(f"💾 Results saved to: {filepath}")

# Usage examples for different platforms
if __name__ == '__main__':
    # Example: Scrape Twitter-like content
    twitter_scraper = SocialMediaScraper('twitter', delay_between_requests=2.0)
    
    # Example: Scrape Reddit-like content  
    reddit_scraper = SocialMediaScraper('reddit', delay_between_requests=1.5)
    
    print("Social Media Scraper modules loaded successfully!")
    print("Usage:")
    print("  scraper = SocialMediaScraper('twitter')")
    print("  posts = scraper.scrape_feed('https://twitter.com/some_user')")
    print("  scraper.save_results(posts)")

File: knowledge_scrapers/social_media_scraper.js
#!/usr/bin/env node
/**
 * Social Media Content Scraper - JavaScript/Node.js Implementation
 * 
 * Optimized for dynamic content scraping with browser automation capabilities.
 * Features:
 * - Puppeteer for JavaScript-heavy sites
 * - Real-time content extraction
 * - Anti-detection measures
 * - Concurrent processing
 */

const puppeteer = require('puppeteer');
const fs = require('fs').promises;
const path = require('path');

class SocialMediaPost {
    constructor(data) {
        this.url = data.url;
        this.platform = data.platform;
        this.author = data.author;
        this.content = data.content;
        this.timestamp = data.timestamp;
        this.likes_count = data.likes_count;
        this.shares_count = data.shares_count;
        this.comments_count = data.comments_count;
        this.hashtags = data.hashtags || [];
        this.scraped_at = new Date().toISOString();
    }
}

class SocialMediaScraperJS {
    constructor(platform, options = {}) {
        this.platform = platform.toLowerCase();
        this.delay = options.delay_between_requests || 2000;
        this.headless = options.headless !== false;
        this.userAgent = 'Marina-SocialMediaScraper/2.0 (Educational Research)';
        this.visitedUrls = new Set();
        
        // Platform-specific configurations optimized for JavaScript execution
        this.platformConfigs = {
            twitter: {
                selectors: {
                    post: '[data-testid="tweet"]',
                    content: '[data-testid="tweetText"]',
                    author: '[data-testid="User-Name"] span',
                    timestamp: 'time',
                    likes: '[data-testid="like"]',
                    shares: '[data-testid="retweet"]',
                    comments: '[data-testid="reply"]'
                },
                waitForSelector: '[data-testid="tweet"]',
                scrollToLoad: true
            },
            reddit: {
                selectors: {
                    post: '[data-testid="post-container"]',
                    content: '[data-test-id="post-content"]',
                    author: '[data-testid="comment_author_link"]',
                    timestamp: '[data-testid="comment_timestamp"]',
                    likes: '[aria-label*="upvote"]',
                    comments: '[data-click-id="comments"]'
                },
                waitForSelector: '[data-testid="post-container"]',
                scrollToLoad: true
            },
            facebook: {
                selectors: {
                    post: '[data-pagelet="FeedUnit"]',
                    content: '[data-testid="post_message"]',
                    author: 'strong a',
                    timestamp: 'abbr',
                    likes: '[aria-label*="Like"]',
                    shares: '[aria-label*="Share"]',
                    comments: '[aria-label*="Comment"]'
                },
                waitForSelector: '[data-pagelet="FeedUnit"]',
                scrollToLoad: true
            }
        };
    }

    async initializeBrowser() {
        console.log('🚀 Initializing browser for social media scraping...');
        
        this.browser = await puppeteer.launch({
            headless: this.headless,
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage',
                '--disable-accelerated-2d-canvas',
                '--no-first-run',
                '--no-zygote',
                '--disable-gpu'
            ]
        });
        
        this.page = await this.browser.newPage();
        
        // Set user agent and viewport
        await this.page.setUserAgent(this.userAgent);
        await this.page.setViewport({ width: 1366, height: 768 });
        
        // Enable request interception for performance
        await this.page.setRequestInterception(true);
        this.page.on('request', (req) => {
            const resourceType = req.resourceType();
            if (['image', 'stylesheet', 'font', 'media'].includes(resourceType)) {
                req.abort();
            } else {
                req.continue();
            }
        });
        
        console.log('✅ Browser initialized successfully');
    }

    extractHashtags(text) {
        const hashtagRegex = /#\w+/g;
        return (text.match(hashtagRegex) || []).map(tag => tag.toLowerCase());
    }

    extractMentions(text) {
        const mentionRegex = /@\w+/g;
        return (text.match(mentionRegex) || []).map(mention => mention.toLowerCase());
    }

    async extractNumberFromText(element, keywords) {
        try {
            const text = await element.textContent();
            for (const keyword of keywords) {
                const patterns = [
                    new RegExp(`(\\d+)\\s*${keyword}`, 'i'),
                    new RegExp(`${keyword}:?\\s*(\\d+)`, 'i'),
                    new RegExp(`${keyword}\\s*\\((\\d+)\\)`, 'i')
                ];
                
                for (const pattern of patterns) {
                    const match = text.match(pattern);
                    if (match) {
                        return parseInt(match[1]);
                    }
                }
            }
        } catch (error) {
            // Silently handle extraction errors
        }
        return null;
    }

    async scrapePost(postElement, feedUrl) {
        try {
            const config = this.platformConfigs[this.platform];
            if (!config) {
                throw new Error(`Unsupported platform: ${this.platform}`);
            }

            // Extract post content with error handling
            const content = await this.safeExtractText(postElement, config.selectors.content) || 'No content available';
            const author = await this.safeExtractText(postElement, config.selectors.author) || 'Unknown';
            
            // Extract timestamp
            let timestamp = null;
            try {
                const timestampElement = await postElement.$(config.selectors.timestamp);
                if (timestampElement) {
                    timestamp = await timestampElement.getAttribute('datetime') || 
                               await timestampElement.textContent();
                }
            } catch (error) {
                // Timestamp extraction failed
            }

            // Extract engagement metrics
            const likesElement = await postElement.$(config.selectors.likes);
            const sharesElement = await postElement.$(config.selectors.shares);
            const commentsElement = await postElement.$(config.selectors.comments);

            const likesCount = likesElement ? await this.extractNumberFromText(likesElement, ['like', 'heart']) : null;
            const sharesCount = sharesElement ? await this.extractNumberFromText(sharesElement, ['share', 'retweet']) : null;
            const commentsCount = commentsElement ? await this.extractNumberFromText(commentsElement, ['comment', 'reply']) : null;

            // Extract hashtags and mentions
            const hashtags = this.extractHashtags(content);
            
            // Generate post URL (simplified)
            const postUrl = `${feedUrl}#post_${Date.now()}`;

            return new SocialMediaPost({
                url: postUrl,
                platform: this.platform,
                author,
                content,
                timestamp,
                likes_count: likesCount,
                shares_count: sharesCount,
                comments_count: commentsCount,
                hashtags
            });

        } catch (error) {
            console.error(`❌ Error scraping post: ${error.message}`);
            return null;
        }
    }

    async safeExtractText(element, selector) {
        try {
            const targetElement = await element.$(selector);
            return targetElement ? await targetElement.textContent() : null;
        } catch (error) {
            return null;
        }
    }

    async scrapeFeed(feedUrl, maxPosts = 10) {
        if (!this.browser) {
            await this.initializeBrowser();
        }

        const posts = [];
        
        try {
            console.log(`🔍 Scraping ${this.platform} feed: ${feedUrl}`);
            
            await this.page.goto(feedUrl, { 
                waitUntil: 'networkidle0',
                timeout: 30000 
            });

            const config = this.platformConfigs[this.platform];
            
            // Wait for posts to load
            try {
                await this.page.waitForSelector(config.waitForSelector, { timeout: 10000 });
            } catch (error) {
                console.warn(`⚠️ Could not find posts selector: ${config.waitForSelector}`);
                return posts;
            }

            // Scroll to load more posts if needed
            if (config.scrollToLoad) {
                await this.autoScroll(maxPosts);
            }

            // Extract posts
            const postElements = await this.page.$$(config.selectors.post);
            console.log(`📊 Found ${postElements.length} post elements`);

            for (let i = 0; i < Math.min(postElements.length, maxPosts); i++) {
                const post = await this.scrapePost(postElements[i], feedUrl);
                if (post) {
                    posts.push(post);
                }
                
                // Add delay between posts
                await this.sleep(this.delay);
            }

            console.log(`✅ Successfully scraped ${posts.length} posts from ${this.platform}`);

        } catch (error) {
            console.error(`❌ Error scraping feed: ${error.message}`);
        }

        return posts;
    }

    async autoScroll(targetPosts) {
        console.log('📜 Auto-scrolling to load more posts...');
        
        let previousHeight = 0;
        let attempts = 0;
        const maxAttempts = 5;

        while (attempts < maxAttempts) {
            // Scroll to bottom
            await this.page.evaluate(() => {
                window.scrollTo(0, document.body.scrollHeight);
            });

            // Wait for new content to load
            await this.sleep(2000);

            const currentHeight = await this.page.evaluate(() => document.body.scrollHeight);
            
            if (currentHeight === previousHeight) {
                attempts++;
            } else {
                attempts = 0;
                previousHeight = currentHeight;
            }

            // Check if we have enough posts
            const config = this.platformConfigs[this.platform];
            const postCount = await this.page.$$(config.selectors.post).then(elements => elements.length);
            
            if (postCount >= targetPosts) {
                break;
            }
        }
        
        console.log('📜 Auto-scrolling completed');
    }

    async sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    async saveResults(posts, filename = null) {
        if (!filename) {
            const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
            filename = `social_media_scrape_${this.platform}_${timestamp}.json`;
        }

        const resultsDir = path.join(__dirname, 'scraping_results');
        
        try {
            await fs.mkdir(resultsDir, { recursive: true });
        } catch (error) {
            // Directory might already exist
        }

        const filepath = path.join(resultsDir, filename);
        
        const results = {
            platform: this.platform,
            total_posts: posts.length,
            scraped_at: new Date().toISOString(),
            posts: posts
        };

        await fs.writeFile(filepath, JSON.stringify(results, null, 2));
        console.log(`💾 Results saved to: ${filepath}`);
    }

    async close() {
        if (this.browser) {
            await this.browser.close();
            console.log('🔒 Browser closed');
        }
    }
}

// CLI interface for standalone usage
async function main() {
    const args = process.argv.slice(2);
    
    if (args.length < 3) {
        console.log('Usage: node social_media_scraper.js <platform> <url> <max_posts>');
        console.log('Example: node social_media_scraper.js twitter https://twitter.com/home 10');
        process.exit(1);
    }

    const [platform, url, maxPosts] = args;
    const scraper = new SocialMediaScraperJS(platform, { headless: true });

    try {
        const posts = await scraper.scrapeFeed(url, parseInt(maxPosts) || 10);
        await scraper.saveResults(posts);
        
        console.log(`\n✅ Scraping completed successfully!`);
        console.log(`📊 Posts scraped: ${posts.length}`);
        
    } catch (error) {
        console.error(`❌ Scraping failed: ${error.message}`);
        process.exit(1);
    } finally {
        await scraper.close();
    }
}

// Export for use in other modules
module.exports = { SocialMediaScraperJS, SocialMediaPost };

// Run CLI if this file is executed directly
if (require.main === module) {
    main();
}
